### ISSUE TITLE
Make email {template tags} more flexible

### ISSUE THREAD

**Issue by mintplugins** [easydigitaldownloads/easy-digital-downloads#4730]
Proposal: Make email template tags more flexible to allow for more specific types of output. 

For example: {download_list} could be used as an <ul> or as inline. Some clients who sell only 1 product in their entire store wouldn't want the <ul>. Other stores which sell many products would likely want the <ul>.

Something like this could allow them to choose how it is output:

{download_list output_type="ul"} OR {download_list output_type="inline"}  

This could then be used by add-ons as well like Software Licensing.

**Comment by pippinsplugins** [comment]
I love this idea so much.

**Comment by cklosowski** [comment]
Agree. Let's build this functionality into core so that extensions don't have to recreate it.

**Comment by danieliser** [comment]
Worked on similar using EDD as a starting point recently for another project.

Came up with an extendable class and function that looks like this:

/**
	 * Processes email notifications.
	 *
	 * $args['code'] string Defines the email notification code used to load templates & content.
	 * $args['merge_vars'] array Defines the default merge variables array for this notification.
	 * $args['do_not_reply'] bool Whether this is a noreply@ email.
	 * $args['allow_user_vars'] bool When personalization is enabled this will trigger user variables to be passed with the merge vars for each subscriber.
	 *
	 * @param array $args
	 * @param string $to
	 * @param array $extra_merge_vars
	 *
	 * @return bool
	 */
	function jpf_process_email_notifications( $args = array(), $to = '', $extra_merge_vars = array() ) {
		$args = wp_parse_args( $args, array(
			'code'            => null,
			'merge_vars'      => array(),
			'do_not_reply'    => true,
			'allow_user_vars' => null,
		) );
	
		$email = jpf_get_email_instance( $args );
	
		if ( ! $email ) {
			return false;
		}
	
		return $email->send( $to, $extra_merge_vars );
	}

Includes caching and looks like this when used

		$merge_vars = apply_filters( 'jpf_new_reply_notification_merge_vars', array(
			'sitename'          => wp_specialchars_decode( get_option( 'blogname' ), ENT_QUOTES ),
			'topic_title'       => $topic->title(),
			'reply_content'     => $reply->content(),
			'reply_author_name' => $reply->author_name(),
			'reply_url'         => $reply->url(),
		) );

		do_action( 'jpf_before_new_reply_notifications', $reply->ID, $topic->ID, $merge_vars, $subscribers );

		jpf_process_email_notifications( array(
			'code'       => 'new_reply',
			'merge_vars' => $merge_vars,
		), $subscribers );

I have a lot of work done in this and it allows many things such as:
- Templating
- Multiple email contents identified by 'code'.
- Caches emails after pre-rendering on first send, merge vars are done just before sending.
- Both predefined sets of merge_vars, as well as ability to pass them in at run time.
- Mail handler is overloadable for use with Amazon SES, MailGun etc.
- Send single emails
- Send batches using either bcc (no personalization) or personalized(using batch send).
- Lots more of optimized goodies in there, but I would have to read through it line by line to point them out.

Its part of a currently private repo, but happy to share if you are interested.

To give you an idea, we are using it for handling all email notifications for a new forum plugin, including personalized subscription notifications, and alert (wordpress.org keyword based notifications), so it is pretty efficient even with 50-100 recipients. If we use Amazon SES we can get even better performance. Something we will likely be including in this premium only plugin.
